ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,899 @@
+{
+    pid_t pid = (pid_t) (intptr_t) arg;
+
+    int status;
+    while (waitpid(pid, &status, 0) < 0 && errno == EINTR)
+        continue;
+
+    return NULL;
+}
+
+static void
+pidwaiter_finalize(value v)
+{
+    PIDWAITER(waiter, v);
+
+    pid_t pid = waiter->pid;
+    bool reaped = waiter->reaped;
+    pthread_mutex_destroy(&waiter->mtx);
+    caml_stat_free(waiter);
+
+    // reap the pid to avoid zombies
+    if (!reaped) {
+        pthread_t th;
+        if (create_thread_minstack(&th, thread_proc_reap, (void *) (intptr_t) pid) == 0)
+            pthread_detach(th);
+    }
+}
+
+static struct custom_operations custom_ops = {
+    "com.xenserver.pidwaiter",
+    pidwaiter_finalize,
+    custom_compare_default,
+    custom_hash_default,
+    custom_serialize_default,
+    custom_deserialize_default,
+    custom_compare_ext_default,
+    custom_fixed_length_default
+};
+
+static value
+pidwaiter_alloc(pid_t pid)
+{
+    // Allocate static memory, we can't retain pointer to
+    // Ocaml dynamic memory when we release Ocaml engine.
+    pidwaiter *waiter = caml_stat_alloc(sizeof(pidwaiter));
+
+    waiter->pid = pid;
+    pthread_mutex_init(&waiter->mtx, NULL);
+    waiter->reaped = false;
+
+    value v = caml_alloc_custom(&custom_ops, sizeof(pidwaiter*), 0, 1);
+    *((pidwaiter**) Data_custom_val(v)) = waiter;
+    return v;
+}
+
+#define log(...) do {} while(0)
+#include "fe_stubs_redirect_algo.h"
+
+typedef struct {
+    char **args;
+    char **environment;
+    mapping *mappings;
+    fd_operation operations[MAX_OPERATIONS];
+    int err;
+    const char *err_func;
+} exec_info;
+
+// Appends a string to *p_dest buffer.
+// It updates *p_dest to point after copied string.
+// Returns copied string.
+static char *
+append_string(char **p_dest, const char *s)
+{
+    char *const dest = *p_dest;
+    size_t const size = strlen(s) + 1;
+    memcpy(dest, s, size);
+    *p_dest = dest + size;
+    return dest;
+}
+
+static void adjust_args(char **args, mapping *const mappings, unsigned num_mappings);
+static void reset_signal_handlers(void);
+static void clear_cgroup(void);
+static const char *get_syslog_key(value syslog_stdout, value args);
+
+CAMLprim value
+caml_safe_exec_nat(value environment,
+    value stdin_fd, value stdout_fd, value stderr_fd,
+    value id_mapping, value syslog_stdout, value redirect, value args)
+{
+    // We need to retain all paramters, they are either array,
+    // list or options.
+    value waiter = Val_unit;
+    CAMLparam5(environment, stdin_fd, stdout_fd, stderr_fd, waiter);
+    CAMLxparam4(id_mapping, syslog_stdout, redirect, args);
+
+    value res;
+    unsigned num_args = 0, num_mappings = 0;
+    const mlsize_t num_environment = caml_array_length(environment);
+    size_t strings_size = 0;
+    const bool redirect_stderr_to_stdout =
+        (syslog_stdout == Val_none || redirect == Val_none) ? false : Bool_val(Some_val(redirect));
+    const char *const key = get_syslog_key(syslog_stdout, args);
```

Comment:
It is not safe to store a pointer to inside an OCaml value, because the garbage collector may move the value, but won't know how to update this pointer.
Instead it needs to be queried each time using `String_val`, etc.
Or make a copy with `caml_copy_string`.

Reply:
See comment above